### ISSUE TITLE
Support RECEPTOR_ANGLE in feed_rotation

### ISSUE THREAD

**Issue by JSKenyon** [ska-sa/codex-africanus#191]
In the general the P Jones matrix (though I believe it is L Jones in the Codex docs) should also be able to account for the RECEPTOR_ANGLE present in the FEED sub-table. In the most basic of cases, RECEPTOR_ANGLE will be `(n_ant, n_receptor)` where `n_receptor == 2`; one angle for X/R and one angle for Y/L. Doing this at present requires a fair amount of manual intervention but it is absolutely necessary for MeerKAT. I will note that this is only possible if the two numbers specified in RECEPTOR_ANGLE are the same (in general they probably should be), as otherwise we are not dealing with a product of rotation matrices (tagging @bennahugo to confirm).

I would suggest that the [`feed_rotation`](https://github.com/ska-sa/codex-africanus/blob/97b6accde38d0f2d44ffa8743e9be8340c0d5501/africanus/rime/dask.py#L61) function accept an additional parameter, e.g. `feed_offsets`, which are added in the appropriate place. This should preserve the current interface, as the default can be zero, but should make constructing P far less fiddly. Something similar is [done in CubiCal](https://github.com/ratt-ru/CubiCal/blob/e49aff3975d8cc91a29a707c2d60036ad883bc8e/cubical/machines/parallactic_machine.py#L117).

**Comment by bennahugo** [comment]
I don't see a case where these are different - that in essence will be non-orthogonality which is not what you physically want when taking a polarimetric measurement.

The -90 is only there to keep meerkat data selfconsistent, in essence X must point up and Y North to compute a parallactic angle in the IEEE convention. In meerkat these are flipped so it is necessary to tag a -90 degree rotation (which flips the signs of Q and V, but this is not really a problem and can be corrected in post processing).

**Comment by JSKenyon** [comment]
I have actually managed to convince myself that this might not be necessary. I believe that this can be accomplished simply by adding one of the values in RECEPTOR_ANGLE to the computed parallactic angles. The reason for this is that I believe that in general the two values in RECEPTOR_ANGLE must be equal or else it cannot be represented as a rotation matrix. If it cannot be represented as a rotation matrix, then it cannot be implemented in this way at all. Additionally, it would cause a slew of other problems if the receptors where not rotated equally as that would introduce non-orthogonality.

**Comment by JSKenyon** [comment]
Thanks @bennahugo - I think I had just arrived at that conclusion. I will check but I think this can actually be done trivially under that assumption.

**Comment by sjperkins** [comment]
Thanks for typing this up. It also clarified confusion in my head w.r.t. the [MSv2.0 FEED specfication](https://casa.nrao.edu/Memos/229.html#SECTION00066000000000000000). In restrospect this is obvious but:

1. Each row represents a feed
2. Each feed can have a different number of receptors (always 2 in the MeerKAT case as pointed out by @JSKenyon)
3. The number of receptors in each row and hence feed is defined by NUM_RECEPTORS.

**Comment by o-smirnov** [comment]
Folks, the data model as specified in the MS allows an individual angle offset per receptor. I agree in practice they'll almost always be identical, but if the data model allows for it and it's easy enough to support it, why not Do The Right Thing(TM) from the start? It costs us very little to drop the "same angle" assumption.

For receptor angles of \phi and \psi, the appropriate matrix is just a rotation matrix (or phase offset matrix, for the circular case), except row 1 uses the angle of receptor 1, and row 2 uses the angle of receptor 2.

**Comment by JSKenyon** [comment]
Ok, then I guess it reverts to my original suggestion and `feed_rotation` needs to have its internals modified such that it can add the two receptor angles in as you describe @o-smirnov. @sjperkins Thoughts?

**Comment by JSKenyon** [comment]
Great @sjperkins!

**Comment by sjperkins** [comment]
> For receptor angles of \phi and \psi, the appropriate matrix is just a rotation matrix (or phase offset matrix, for the circular case), except row 1 uses the angle of receptor 1, and row 2 uses the angle of receptor 2.

Note that each feed can have a variable number of receptors (not just 2 for the MeerKAT case). Does this imply a composition of rotation matrices?

i.e. given angles rho, phi, psi, is the feed rotation the following: ``dot(rot(rho), dot(rot(phi), dot(psi)))``?

**Comment by JSKenyon** [comment]
It isn't that simple I believe - if it was, my idea regarding simple addition would work. If they were all pure rotation matrices, we could invoke `rot(phi + psi) = rot(phi) * rot(psi)`. I believe that @o-smirnov  is suggesting this in the linear case: 
```
[[cos(pa + psi_1), -sin(pa + psi_1)], 
 [sin(pa + psi_2),  cos(pa + psi_2)]]
``` 
where `psi_1` is the receptor angle for X/R and `psi_2` is the receptor angle for Y/L. I don't know if this can be decomposed in a simple way. 

For circular feeds it is easier and would be:
```
[[e**i(pa + psi_1), 0], 
 [0, e**-i(pa + psi_2)]]
``` 
@o-smirnov could you please confirm that I am not talking nonsense?

**Comment by o-smirnov** [comment]
@JSKenyon correct, modulo the +/- signs in the exponent, the order of which I always forget...

**Comment by bennahugo** [comment]
See the Hales paper I believe the matrices are all given or just copy what we have in cubical v1.

**Comment by sjperkins** [comment]
Pseudocode based on our discussion @JSKenyon 

```python
utime = np.unique(time)
ufeed = np.unique(np.concatenate([feed1, feed2]))
uantenna = np.unique(np.concatenate([antenna1, antenna1]))

parangles = np.zeros((utime, ufeed, uantenna, 2))

feed_receptors = np.zeros(nrows_in_feed)

for t in range(len(range(utime))):
    for f, feed_id in enumerate(ufeed):
        nreceptors = feed_receptors[feed_id]
        assert nreceptors == 2, "Who doesn't have 2 feed receptors?"
        receptor_angles = feed_receptors[feed_id]
        assert len(receptor_angles) == nreceptors 

        for a in range(len(range(uantenna))):
            parangles[t, f, a, 0] = compute(..., receptor_angles[0])
            parangles[t, f, a, 1] = compute(..., receptor_angles[1])
```

**Comment by JSKenyon** [comment]
I have been tackling this on the QuartiCal end. Specifically, I have been working to better understand what happens when we have non-orthogonal receptors. This would be highly unusual but is not entirely impossible. References/derivations for this case are thin on the ground. However, I found [this link](https://casa.nrao.edu/aips2_docs/notes/185/node6.html) to be incredibly useful. I have used it to (re-)derive the following general expressions for P jones, where psi is the parallactic angle and theta is the angle associated with a specific receptor (given by the subscript). 

Linear feeds:
![nonorth_l](https://user-images.githubusercontent.com/6582745/140738923-69722579-9c26-460f-9989-c826f45c3bf3.png)

Circular feeds:
![nonorth](https://user-images.githubusercontent.com/6582745/140739185-4037ae41-9e79-4aeb-b580-be96a7ce1148.png)

Note that there may be slight sign discrepancies. I am still endeavouring to sort them out. The signs used by [Hales](https://arxiv.org/pdf/1706.06612.pdf) are the exact opposite of those provided in the link above.

**Comment by sjperkins** [comment]
https://github.com/ska-sa/codex-africanus/commit/4d74742afe2e97b665550754c8b87fe8906499d4 implements the feed rotation as per the equations above.

**Comment by sjperkins** [comment]
> Linear feeds: ![nonorth_l](https://user-images.githubusercontent.com/6582745/140738923-69722579-9c26-460f-9989-c826f45c3bf3.png)
> 
> Circular feeds: ![nonorth](https://user-images.githubusercontent.com/6582745/140739185-4037ae41-9e79-4aeb-b580-be96a7ce1148.png)


Note that I ended up implementing the reverse sign of these expressions.

Closed by #239.
